test/NameBinding/dynamic-callable.swift

```diff
@@ -0,0 +1,373 @@
+    case .multiply:
+      return x * y
+    }
+  }
+}
+
+func testEnum() {
+  let ops: [BinaryOperation<Int>] = [.add, .subtract, .multiply]
+  for op in ops {
+    print(op(3, 4))
+  }
+}
+
+//===----------------------------------------------------------------------===//
+// Generics
+//===----------------------------------------------------------------------===//
+
+@dynamicCallable
+struct CallableGenericArray<A : ExpressibleByArrayLiteral> {
+  func dynamicallyCall(withArguments arguments: A) -> Int {
+    return 1
+  }
+}
+func testGenericArray<A : ExpressibleByArrayLiteral>(
+  a: CallableGenericArray<A>, x: A.ArrayLiteralElement
+) -> Int {
+  return a() + a(x, x)
+}
+
+@dynamicCallable
+struct CallableGenericDictionary<D : ExpressibleByDictionaryLiteral>
+  where D.Key : ExpressibleByStringLiteral {
+  func dynamicallyCall(withKeywordArguments arguments: D) -> Int {
+    return 1
+  }
+}
+func testGenericDictionary<D : ExpressibleByDictionaryLiteral>(
+  a: CallableGenericDictionary<D>, x: D.Value
+) -> Int where D.Key : ExpressibleByStringLiteral {
+  return a() + a(label1: x, x, label2: x)
+}
+
+@dynamicCallable
+struct CallableGeneric1<T> {
+  func dynamicallyCall(withArguments arguments: [T]) -> Int {
+    return arguments.count
+  }
+}
+func testGenericType1<T>(a: CallableGeneric1<T>, x: T) -> Int {
+  return a() + a(x, x)
+}
+func testConcreteGenericType2(a: CallableGeneric1<Int>) -> Int {
+  return a() + a(1, 2)
+}
+
+@dynamicCallable
+struct CallableGeneric2<T> {
+  func dynamicallyCall(withArguments arguments: [Any]) -> Int {
+    return arguments.count
+  }
+}
+func testGenericType2<T>(a: CallableGeneric2<T>) -> Int {
+  return a(1, 2) + a("asdf", 123)
+}
+func testConcreteGenericType2(a: CallableGeneric2<Int>) -> Int {
+  return a(1, 2) + a("asdf", 123)
+}
+
+@dynamicCallable
+struct CallableGeneric3<T> {
+  func dynamicallyCall(
+    withKeywordArguments arguments: DictionaryLiteral<String, T>
+  ) -> Int {
+    return arguments.count
+  }
+}
+func testGenericType3<T>(a: CallableGeneric3<T>, x: T) -> Int {
+  return a() + a(x1: x, x, x, x2: x)
+}
+func testConcreteGenericType3(a: CallableGeneric3<Int>) -> Int {
+  return a() + a(x1: 123, 1, 2, x2: 123)
+}
+
+@dynamicCallable
+struct CallableGeneric4<T> {
+  func dynamicallyCall<U>(withArguments arguments: [U]) -> Int {
+    return arguments.count
+  }
+
+  func dynamicallyCall<U>(
+    withKeywordArguments arguments: DictionaryLiteral<StaticString, U>
+  ) -> Int {
+    return arguments.count
+  }
+}
+func testGenericType4<T>(a: CallableGeneric4<T>) -> Int {
+  return a() + a(1, 2, 3) + a(x1: 1, 2, x3: 3)
+}
+
+@dynamicCallable
+struct CallableGeneric5<T> {
+  func dynamicallyCall<U>(withArguments arguments: [U]) -> U {
+    return arguments[0]
+  }
+
+  func dynamicallyCall<U>(
+    withKeywordArguments arguments: DictionaryLiteral<StaticString, U>
+  ) -> U {
+    return arguments[0].1
+  }
+}
+func testGenericType5<T>(a: CallableGeneric5<T>) -> Double {
```

Comment:
Please also test dynamic calls on an archetype:

```
func f<T>(_ t: T) where T : ... {
  t.foo()
}
```

With protocol and class constraints on `T`.

Reply:
Sorry for the late response.

---
I added [such a test](https://github.com/apple/swift/pull/16980/commits/fe91f19de1c273e6179aaa08a71fa145ae314d41#diff-e215fa74b04a179ecc272f8d774d7199R382) and it failed - thank you for the catch!

I'll take a closer look this weekend. It seems `@dynamicMemberLookup` has the same issue with archetypes:
```swift
@dynamicMemberLookup
class MemberLookup {
  subscript(dynamicMember name: String) -> Int { return 1 }
}

func testPass(a: MemberLookup) -> Int {
  return a.foo // ok
}
func testFail<T : MemberLookup>(a: T) -> Int {
  return a.foo // error: value of type 'T' has no member 'foo'
}
```